### PR TITLE
Fix VRF HeatRecovery Check

### LINKED ISSUE #10514: VRF Heat Recovery Mistakenly Disabled
Issue overview
--------------
From user:
In the attached model, there are 3 VRF outdoor units
* HP-1: serves multiple terminal units + heat recovery
* HP-2: serves multiple terminal units + heat recovery
* HP-3: serves single terminal units with no heat recovery
There are a lot of unmet load hours because the heat recovery on HP-1 is disabled. I see the heating went to zero when there is cooling, and there is no "VRF Heat Pump Heat Recovery Status Change Multiplier" output for HP-1.

I suspect the heat recovery of HP-1 is disabled because HP-3 because I see a warning message below, and once I remove HP-3 from the model OR connect one more terminal unit to the HP-3 the UMLH is reduced and "Heat Recovery Status" is reported for HP-1.
```
   ** Warning ** ZoneTerminalUnitList "HP-3 TERMINAL LIST"
   ** Warning ** ...Only 1 Terminal Unit connected to system and heat recovery is selected.
   ** ~~~ ** ...Heat recovery will be disabled.
```
*************

There are several things wrong with the loop that checks for this. With the current code, if any terminal unit list has only one terminal unit, and the system connected to the first terminal unit in the idf file has heat recovery, then the warning will be written and heat recovery for the first AirConditioner:VariableRefrigerantFlow in the idf file will be disabled.

https://github.com/NREL/EnergyPlus/blob/281f7a4bfa9399f188323fbc94b14445d21ee4da/src/EnergyPlus/HVACVariableRefrigerantFlow.cc#L4618-L4644

### Workaround
The workaround is to rearrange the AirConditioner:VariableRefrigerantFlow and terminal unit objects in the idf file. The ZoneHVAC:TerminalUnit:VariableRefrigerantFlow and AirConditioner:VariableRefrigerantFlow without heat recovery should be the first object of its respective type.

### Details
Some additional details for this issue (if relevant):
 - Version of EnergyPlus 24.2-9d7789a3ac
 - Helpdesk ticket number 16661.

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added EnergyPlusDevSupport\DefectFiles...
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by mjwitte**
Pull request overview
---------------------
 - Fixes #10514 

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [x] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 1]
```diff
@@ -4616,27 +4616,28 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
...
-                                  "...Zone Terminal Unit = " + state.dataHVACVarRefFlow->TerminalUnitList(NumList).ZoneTUName(VRFTUNum) +
-                                      " improperly connected to system.");
+                ShowSevereError(state, format("ZoneTerminalUnitList \"{}\"", thisTUList.Name));
+                ShowContinueError(state, format("...Zone Terminal Unit = {} improperly connected to system.", thisTUList.ZoneTUName(VRFTUNum)));
                 ShowContinueError(state, "...either the ZoneHVAC:TerminalUnit:VariableRefrigerantFlow object does not exist,");
                 ShowContinueError(state, "...the ZoneHVAC:TerminalUnit:VariableRefrigerantFlow object name is misspelled,");
                 ShowContinueError(state, "...or the ZoneTerminalUnitList object is not named in an AirConditioner:VariableRefrigerantFlow object.");
                 ErrorsFound = true;
-            }
-            if (state.dataHVACVarRefFlow->VRFTU(VRFTUNum).VRFSysNum > 0) {
```
Comment (mjwitte) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc:4632]: Here `VRFTUNum` should have been `state.dataHVACVarRefFlow->TerminalUnitList(NumList).ZoneTUPtr(VRFTUNum)`.

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 2]
```diff
@@ -4616,27 +4616,28 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
...
-                if (state.dataHVACVarRefFlow->TerminalUnitList(NumList).NumTUInList == 1 &&
-                    state.dataHVACVarRefFlow->VRF(state.dataHVACVarRefFlow->VRFTU(VRFTUNum).VRFSysNum).VRFAlgorithmType == AlgorithmType::SysCurve) {
-                    if (state.dataHVACVarRefFlow->VRF(state.dataHVACVarRefFlow->VRFTU(VRFTUNum).VRFSysNum).HeatRecoveryUsed) {
```
Comment (mjwitte) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc:4635]: Same thing again here.

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 3]
```diff
@@ -4616,27 +4616,28 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
...
-                        ShowWarningError(state, "ZoneTerminalUnitList \"" + state.dataHVACVarRefFlow->TerminalUnitList(NumList).Name + "\"");
-                        ShowWarningError(state, "...Only 1 Terminal Unit connected to system and heat recovery is selected.");
-                        ShowContinueError(state, "...Heat recovery will be disabled.");
-                        state.dataHVACVarRefFlow->VRF(VRFTUNum).HeatRecoveryUsed = false;
```
Comment (mjwitte) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc:4639]: And here `VRFTUNum` should have been `state.dataHVACVarRefFlow->VRFTU(state.dataHVACVarRefFlow->TerminalUnitList(NumList).ZoneTUPtr(VRFTUNum)).VRFSysNum``.

**Comment by mjwitte** [comment]
Defect file, before:
```
   ** Warning ** ZoneTerminalUnitList "HP-3 TERMINAL LIST"
   ** Warning ** ...Only 1 Terminal Unit connected to system and heat recovery is selected.
```
image

With this branch:
No error message about heat recovery disabled, and both HP1 and 2 have heat recovery status output.
image

**Review by mjwitte — COMMENTED** [review]
Code walkthru. Probably easiest to review by looking at the new code carefully without diffs.

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 4]
```diff
@@ -4616,27 +4616,28 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
     } // end Number of VRF Terminal Unit Loop
 
     //   perform additional error checking
-    for (int NumList = 1; NumList <= state.dataHVACVarRefFlow->NumVRFTULists; ++NumList) {
-        for (int VRFTUNum = 1; VRFTUNum <= state.dataHVACVarRefFlow->TerminalUnitList(NumList).NumTUInList; ++VRFTUNum) {
-            if (state.dataHVACVarRefFlow->TerminalUnitList(NumList).ZoneTUPtr(VRFTUNum) == 0) {
+    for (auto const &thisTUList : state.dataHVACVarRefFlow->TerminalUnitList) {
+        for (int VRFTUNum = 1; VRFTUNum <= thisTUList.NumTUInList; ++VRFTUNum) {
+            int const tuPtr = thisTUList.ZoneTUPtr(VRFTUNum);
+            if (tuPtr == 0) {
```
Comment (rraustad) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc:4622]: Yes, does need to point to the correct TU, not the index in the list. The change here is the same as before (i.e., it did point to the correct TU), but below at old line 4632 is the problem and this change corrects that.

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 5]
```diff
@@ -4616,27 +4616,28 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
...
-            }
-            if (state.dataHVACVarRefFlow->VRFTU(VRFTUNum).VRFSysNum > 0) {
-                if (state.dataHVACVarRefFlow->TerminalUnitList(NumList).NumTUInList == 1 &&
-                    state.dataHVACVarRefFlow->VRF(state.dataHVACVarRefFlow->VRFTU(VRFTUNum).VRFSysNum).VRFAlgorithmType == AlgorithmType::SysCurve) {
-                    if (state.dataHVACVarRefFlow->VRF(state.dataHVACVarRefFlow->VRFTU(VRFTUNum).VRFSysNum).HeatRecoveryUsed) {
-                        ShowWarningError(state, "ZoneTerminalUnitList \"" + state.dataHVACVarRefFlow->TerminalUnitList(NumList).Name + "\"");
-                        ShowWarningError(state, "...Only 1 Terminal Unit connected to system and heat recovery is selected.");
-                        ShowContinueError(state, "...Heat recovery will be disabled.");
-                        state.dataHVACVarRefFlow->VRF(VRFTUNum).HeatRecoveryUsed = false;
+            } else {
```
Comment (rraustad) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc:4630]: The use of the else here does seem to work as expected. I question the SysCurve conditional below (4634) and wonder what check if any is needed for the FluidTCtrl model.

File: tst/EnergyPlus/unit/HVACVariableRefrigerantFlow.unit.cc [hunk 6]
```diff
@@ -24251,4 +24251,1185 @@ TEST_F(EnergyPlusFixture, VRF_MultispeedFan_Test)
...
+    // VRF1 serves 2 TUs and Heat Pump Waste Heat Recovery=Yes, expect true
+    int const vrf1 = Util::FindItemInList("VRF OUTDOOR UNIT 1", state->dataHVACVarRefFlow->VRF);
+    EXPECT_TRUE(state->dataHVACVarRefFlow->VRF(vrf1).HeatRecoveryUsed);
+    // VRF2 serves 1 TU and Heat Pump Waste Heat Recovery=No, expect false
+    int const vrf2 = Util::FindItemInList("VRF OUTDOOR UNIT 2", state->dataHVACVarRefFlow->VRF);
+    EXPECT_FALSE(state->dataHVACVarRefFlow->VRF(vrf2).HeatRecoveryUsed);
+    // VRF3 serves 1 TU and Heat Pump Waste Heat Recovery=Yes, expect heat recovery to be disabled, false
+    int const vrf3 = Util::FindItemInList("VRF OUTDOOR UNIT 3", state->dataHVACVarRefFlow->VRF);
+    EXPECT_FALSE(state->dataHVACVarRefFlow->VRF(vrf3).HeatRecoveryUsed);
+}
```
Comment (rraustad) [inline tst/EnergyPlus/unit/HVACVariableRefrigerantFlow.unit.cc:25433]: Unit test covers the issue. The substring compare makes this much cleaner.

**Review by rraustad — APPROVED** [review]
Looks good.

**Comment by Myoldmopar** [comment]
I pulled develop in and it's all still happy, merging this.  Thanks @mjwitte and @rraustad

**Pull request merged**